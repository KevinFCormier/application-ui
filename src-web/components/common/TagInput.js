/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2018. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 *******************************************************************************/

import React from 'react'
import PropTypes from 'prop-types'
import ReactTags from 'react-tag-autocomplete'
import resources from '../../../lib/shared/resources'
import { Icon } from 'carbon-components-react'
import msgs from '../../../nls/platform.properties'
// import FilterModal from '../modals/FilterModal'

resources(() => {
  require('../../../scss/reactTag.scss')
})

class TagInput extends React.Component {
  constructor(props) {
    super(props)
    this.handleDelete = this.handleDelete.bind(this)
    this.handleAddition = this.handleAddition.bind(this)
    this.handleDrag = this.handleDrag.bind(this)
    this.handleClearAllClick = this.handleClearAllClick.bind(this)
  }

  componentWillMount() {
    const { tags=[] } = this.props
    this.setState({
      tags: tags,
    })
  }

  updateSelectedTags(tags) {
    const { onSelectedFilterChange } = this.props
    this.setState({ tags: tags })
    onSelectedFilterChange && onSelectedFilterChange(tags)
  }

  handleDelete(i) {
    const { tags } = this.state
    this.updateSelectedTags(tags.filter((tag, index) => index !== i))
  }

  handleAddition(tag) {
    const { availableFilters=[] } = this.props
    const suggestions = this.convertObjectToArray(availableFilters)
    if (!tag.type) {
      // user can only add a tag that exists in suggestions
      tag = suggestions.find(element => element.name === tag.name)
    }
    if (tag) this.updateSelectedTags([...this.state.tags, tag], tag)
  }

  handleDrag(tag, currPos, newPos) {
    const tags = [...this.state.tags]
    const newTags = tags.slice()

    newTags.splice(currPos, 1)
    newTags.splice(newPos, 0, tag)

    this.updateSelectedTags(newTags)
  }

  onFilterButtonClick() {
    this.handleModalOpen()
  }

  handleClearAllClick() {
    this.updateSelectedTags([])
  }

  handleModalOpen = () => {
    this.setState((prevState) => Object.assign({}, prevState, {
      modalOpen: true
    }))
  }

  handleModalClose = () => {
    this.setState((prevState) => Object.assign({}, prevState, {
      modalOpen: true
    }))
  }

  convertObjectToArray(input) {
    if (Array.isArray(input)) {
      return input
    } else {
      let result = []
      Object.values(input).forEach(value => {
        if (Array.isArray(value)) {
          Object.values(value).forEach(element => {
            result = [...result, element]
          })
        }
      })
      return result
    }
  }

  render() {
    const { availableFilters=[], onFilterButtonClick=this.onFilterButtonClick, hideModalButton } = this.props
    const { tags } = this.state
    const suggestions = this.convertObjectToArray(availableFilters)
    return (
      <div className='tagInput-filter'>
        {!hideModalButton &&
          <div className='tagInput-filterButton'>
            <Icon
              className='icon--filter--glyph'
              name='icon--filter--glyph'
              description={'filter'}
              onClick={onFilterButtonClick} />
          </div>
        }
        {/*<FilterModal selected={tags}*/}
        {/*suggestions={suggestions}*/}
        {/*modalOpen={this.state.modalOpen}*/}
        {/*handleModalOpen={this.handleModalOpen()}*/}
        {/*handleModalClose={this.handleModalClose()}*/}
        {/*/>*/}
        <div className='tagInput-comboBox'>
          <ReactTags
            placeholder={msgs.get('placeholder.filterInput.tags', this.context.locale)}
            tags={tags}
            suggestions={suggestions}
            handleDelete={this.handleDelete}
            handleAddition={this.handleAddition}
            autoresize={false}
            minQueryLength={1}
          />
        </div>
        <div role='presentation' className='tagInput-clearAll' onClick={this.handleClearAllClick}>{msgs.get('actions.clearAll', this.context.locale)}</div>
      </div>
    )
  }
}

TagInput.propTypes = {
  availableFilters: PropTypes.oneOfType([PropTypes.array, PropTypes.object]),
  hideModalButton: PropTypes.bool,
  onFilterButtonClick: PropTypes.func,
  onSelectedFilterChange: PropTypes.func,
  tags: PropTypes.array,
}

export default TagInput
