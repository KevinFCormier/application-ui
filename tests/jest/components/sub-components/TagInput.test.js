/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2018. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 *******************************************************************************/
'use strict'

import React from 'react'
import { shallow } from 'enzyme'

import TagInput from '../../../../src-web/components/common/TagInput'

describe('TagInput component 1', () => {
  const fn = jest.fn()
  const tagFilterProps = {
    tags: [
      { id: 'tag1', name: 'tag1' },
      { id: 'tag2', name: 'tag2' }
    ],
    suggestions: [
      { id: 'tag1', name: 'tag1' },
      { id: 'tag2', name: 'tag2' },
      { id: 'tag3', name: 'tag3' },
      { id: 'tag4', name: 'tag4' },
      { id: 'tag5', name: 'tag5' },
      { id: 'tag6', name: 'tag6' }
    ]
  }
  it('renders as expected', () => {
    const component = shallow(
      <TagInput  {...tagFilterProps} hideModalButton={true} onFilterButtonClick={fn} />
    )
    expect(component).toMatchSnapshot()
    //clear all out button
    component.find('.tagInput-clearAll').at(0).simulate('click')
    expect(component).toMatchSnapshot()
  })
})
