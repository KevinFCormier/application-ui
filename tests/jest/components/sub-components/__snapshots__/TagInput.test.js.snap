// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TagInput component 1 renders as expected 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <TagInput
    hideModalButton={true}
    onFilterButtonClick={[MockFunction]}
    suggestions={
      Array [
        Object {
          "id": "tag1",
          "name": "tag1",
        },
        Object {
          "id": "tag2",
          "name": "tag2",
        },
        Object {
          "id": "tag3",
          "name": "tag3",
        },
        Object {
          "id": "tag4",
          "name": "tag4",
        },
        Object {
          "id": "tag5",
          "name": "tag5",
        },
        Object {
          "id": "tag6",
          "name": "tag6",
        },
      ]
    }
    tags={
      Array [
        Object {
          "id": "tag1",
          "name": "tag1",
        },
        Object {
          "id": "tag2",
          "name": "tag2",
        },
      ]
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        false,
        <div
          className="tagInput-comboBox"
        >
          <ReactTags
            allowBackspace={true}
            allowNew={false}
            autofocus={true}
            autoresize={false}
            delimiterChars={Array []}
            delimiters={
              Array [
                9,
                13,
              ]
            }
            handleAddition={[Function]}
            handleDelete={[Function]}
            maxSuggestionsLength={6}
            minQueryLength={1}
            placeholder="Add new filter tags..."
            suggestions={Array []}
            tagComponent={null}
            tags={
              Array [
                Object {
                  "id": "tag1",
                  "name": "tag1",
                },
                Object {
                  "id": "tag2",
                  "name": "tag2",
                },
              ]
            }
          />
        </div>,
        <div
          className="tagInput-clearAll"
          onClick={[Function]}
          role="presentation"
        >
          Clear All
        </div>,
      ],
      "className": "tagInput-filter",
    },
    "ref": null,
    "rendered": Array [
      false,
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <ReactTags
            allowBackspace={true}
            allowNew={false}
            autofocus={true}
            autoresize={false}
            delimiterChars={Array []}
            delimiters={
              Array [
                9,
                13,
              ]
            }
            handleAddition={[Function]}
            handleDelete={[Function]}
            maxSuggestionsLength={6}
            minQueryLength={1}
            placeholder="Add new filter tags..."
            suggestions={Array []}
            tagComponent={null}
            tags={
              Array [
                Object {
                  "id": "tag1",
                  "name": "tag1",
                },
                Object {
                  "id": "tag2",
                  "name": "tag2",
                },
              ]
            }
          />,
          "className": "tagInput-comboBox",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "allowBackspace": true,
            "allowNew": false,
            "autofocus": true,
            "autoresize": false,
            "delimiterChars": Array [],
            "delimiters": Array [
              9,
              13,
            ],
            "handleAddition": [Function],
            "handleDelete": [Function],
            "maxSuggestionsLength": 6,
            "minQueryLength": 1,
            "placeholder": "Add new filter tags...",
            "suggestions": Array [],
            "tagComponent": null,
            "tags": Array [
              Object {
                "id": "tag1",
                "name": "tag1",
              },
              Object {
                "id": "tag2",
                "name": "tag2",
              },
            ],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Clear All",
          "className": "tagInput-clearAll",
          "onClick": [Function],
          "role": "presentation",
        },
        "ref": null,
        "rendered": "Clear All",
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          false,
          <div
            className="tagInput-comboBox"
          >
            <ReactTags
              allowBackspace={true}
              allowNew={false}
              autofocus={true}
              autoresize={false}
              delimiterChars={Array []}
              delimiters={
                Array [
                  9,
                  13,
                ]
              }
              handleAddition={[Function]}
              handleDelete={[Function]}
              maxSuggestionsLength={6}
              minQueryLength={1}
              placeholder="Add new filter tags..."
              suggestions={Array []}
              tagComponent={null}
              tags={
                Array [
                  Object {
                    "id": "tag1",
                    "name": "tag1",
                  },
                  Object {
                    "id": "tag2",
                    "name": "tag2",
                  },
                ]
              }
            />
          </div>,
          <div
            className="tagInput-clearAll"
            onClick={[Function]}
            role="presentation"
          >
            Clear All
          </div>,
        ],
        "className": "tagInput-filter",
      },
      "ref": null,
      "rendered": Array [
        false,
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <ReactTags
              allowBackspace={true}
              allowNew={false}
              autofocus={true}
              autoresize={false}
              delimiterChars={Array []}
              delimiters={
                Array [
                  9,
                  13,
                ]
              }
              handleAddition={[Function]}
              handleDelete={[Function]}
              maxSuggestionsLength={6}
              minQueryLength={1}
              placeholder="Add new filter tags..."
              suggestions={Array []}
              tagComponent={null}
              tags={
                Array [
                  Object {
                    "id": "tag1",
                    "name": "tag1",
                  },
                  Object {
                    "id": "tag2",
                    "name": "tag2",
                  },
                ]
              }
            />,
            "className": "tagInput-comboBox",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "allowBackspace": true,
              "allowNew": false,
              "autofocus": true,
              "autoresize": false,
              "delimiterChars": Array [],
              "delimiters": Array [
                9,
                13,
              ],
              "handleAddition": [Function],
              "handleDelete": [Function],
              "maxSuggestionsLength": 6,
              "minQueryLength": 1,
              "placeholder": "Add new filter tags...",
              "suggestions": Array [],
              "tagComponent": null,
              "tags": Array [
                Object {
                  "id": "tag1",
                  "name": "tag1",
                },
                Object {
                  "id": "tag2",
                  "name": "tag2",
                },
              ],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Clear All",
            "className": "tagInput-clearAll",
            "onClick": [Function],
            "role": "presentation",
          },
          "ref": null,
          "rendered": "Clear All",
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`TagInput component 1 renders as expected 2`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <TagInput
    hideModalButton={true}
    onFilterButtonClick={[MockFunction]}
    suggestions={
      Array [
        Object {
          "id": "tag1",
          "name": "tag1",
        },
        Object {
          "id": "tag2",
          "name": "tag2",
        },
        Object {
          "id": "tag3",
          "name": "tag3",
        },
        Object {
          "id": "tag4",
          "name": "tag4",
        },
        Object {
          "id": "tag5",
          "name": "tag5",
        },
        Object {
          "id": "tag6",
          "name": "tag6",
        },
      ]
    }
    tags={
      Array [
        Object {
          "id": "tag1",
          "name": "tag1",
        },
        Object {
          "id": "tag2",
          "name": "tag2",
        },
      ]
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        false,
        <div
          className="tagInput-comboBox"
        >
          <ReactTags
            allowBackspace={true}
            allowNew={false}
            autofocus={true}
            autoresize={false}
            delimiterChars={Array []}
            delimiters={
              Array [
                9,
                13,
              ]
            }
            handleAddition={[Function]}
            handleDelete={[Function]}
            maxSuggestionsLength={6}
            minQueryLength={1}
            placeholder="Add new filter tags..."
            suggestions={Array []}
            tagComponent={null}
            tags={Array []}
          />
        </div>,
        <div
          className="tagInput-clearAll"
          onClick={[Function]}
          role="presentation"
        >
          Clear All
        </div>,
      ],
      "className": "tagInput-filter",
    },
    "ref": null,
    "rendered": Array [
      false,
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <ReactTags
            allowBackspace={true}
            allowNew={false}
            autofocus={true}
            autoresize={false}
            delimiterChars={Array []}
            delimiters={
              Array [
                9,
                13,
              ]
            }
            handleAddition={[Function]}
            handleDelete={[Function]}
            maxSuggestionsLength={6}
            minQueryLength={1}
            placeholder="Add new filter tags..."
            suggestions={Array []}
            tagComponent={null}
            tags={Array []}
          />,
          "className": "tagInput-comboBox",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "allowBackspace": true,
            "allowNew": false,
            "autofocus": true,
            "autoresize": false,
            "delimiterChars": Array [],
            "delimiters": Array [
              9,
              13,
            ],
            "handleAddition": [Function],
            "handleDelete": [Function],
            "maxSuggestionsLength": 6,
            "minQueryLength": 1,
            "placeholder": "Add new filter tags...",
            "suggestions": Array [],
            "tagComponent": null,
            "tags": Array [],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Clear All",
          "className": "tagInput-clearAll",
          "onClick": [Function],
          "role": "presentation",
        },
        "ref": null,
        "rendered": "Clear All",
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          false,
          <div
            className="tagInput-comboBox"
          >
            <ReactTags
              allowBackspace={true}
              allowNew={false}
              autofocus={true}
              autoresize={false}
              delimiterChars={Array []}
              delimiters={
                Array [
                  9,
                  13,
                ]
              }
              handleAddition={[Function]}
              handleDelete={[Function]}
              maxSuggestionsLength={6}
              minQueryLength={1}
              placeholder="Add new filter tags..."
              suggestions={Array []}
              tagComponent={null}
              tags={Array []}
            />
          </div>,
          <div
            className="tagInput-clearAll"
            onClick={[Function]}
            role="presentation"
          >
            Clear All
          </div>,
        ],
        "className": "tagInput-filter",
      },
      "ref": null,
      "rendered": Array [
        false,
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <ReactTags
              allowBackspace={true}
              allowNew={false}
              autofocus={true}
              autoresize={false}
              delimiterChars={Array []}
              delimiters={
                Array [
                  9,
                  13,
                ]
              }
              handleAddition={[Function]}
              handleDelete={[Function]}
              maxSuggestionsLength={6}
              minQueryLength={1}
              placeholder="Add new filter tags..."
              suggestions={Array []}
              tagComponent={null}
              tags={Array []}
            />,
            "className": "tagInput-comboBox",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "allowBackspace": true,
              "allowNew": false,
              "autofocus": true,
              "autoresize": false,
              "delimiterChars": Array [],
              "delimiters": Array [
                9,
                13,
              ],
              "handleAddition": [Function],
              "handleDelete": [Function],
              "maxSuggestionsLength": 6,
              "minQueryLength": 1,
              "placeholder": "Add new filter tags...",
              "suggestions": Array [],
              "tagComponent": null,
              "tags": Array [],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Clear All",
            "className": "tagInput-clearAll",
            "onClick": [Function],
            "role": "presentation",
          },
          "ref": null,
          "rendered": "Clear All",
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
